### PR TITLE
Improve serialization of ScriptInstance

### PR DESCRIPTION
**Pull request by orefkov**
Added serialization of Array<String> and Array<Variant> attributes of ScriptInstance.
Added method Swap to script type Array<T> for swap content of two arrays to avoid copying.

### CODE REVIEW

**Comment by eugeneko** [comment]
This makes sence, I'll try it later. Variant[] and String[] will be automatically serializable, am I right?

File: Source/Urho3D/AngelScript/Addons.cpp [hunk 1]
```diff
@@ -1112,6 +1112,18 @@ int CScriptArray::FindByRef(asUINT startAt, void *ref) const
     return -1;
 }
 
+bool CScriptArray::Swap(CScriptArray& other)
+{
+    if (&other != this && other.GetArrayObjectType() == GetArrayObjectType())
+    {
+        SArrayBuffer* t = buffer;
```
Comment (eugeneko) [inline Source/Urho3D/AngelScript/Addons.cpp]: `Urho3D::Swap(t, buffer);` is better, I think.
Also I think that `&other != this` is redundant.

File: Source/Urho3D/AngelScript/Addons.cpp [hunk 2]
```diff
@@ -1647,7 +1659,8 @@ void RegisterArray(asIScriptEngine* engine)
     engine->RegisterObjectMethod("Array<T>", "int Find(uint, const T&in) const", asMETHODPR(CScriptArray, Find, (asUINT, void*) const, int), asCALL_THISCALL);
     engine->RegisterObjectMethod("Array<T>", "int FindByRef(const T&in) const", asMETHODPR(CScriptArray, FindByRef, (void*) const, int), asCALL_THISCALL);
     engine->RegisterObjectMethod("Array<T>", "int FindByRef(uint, const T&in) const", asMETHODPR(CScriptArray, FindByRef, (asUINT, void*) const, int), asCALL_THISCALL);
-    engine->RegisterObjectMethod("Array<T>", "bool opEquals(const Array<T>&in) const", asMETHOD(CScriptArray, operator==), asCALL_THISCALL);
+    engine->RegisterObjectMethod("Array<T>", "bool Swap(Array<T>&in)", asMETHOD(CScriptArray, Swap), asCALL_THISCALL);
+	engine->RegisterObjectMethod("Array<T>", "bool opEquals(const Array<T>&in) const", asMETHOD(CScriptArray, operator==), asCALL_THISCALL);
```
Comment (eugeneko) [inline Source/Urho3D/AngelScript/Addons.cpp]: Tab to space

File: Source/Urho3D/AngelScript/ScriptInstance.cpp [hunk 3]
```diff
@@ -642,6 +675,10 @@ void ScriptInstance::GetScriptAttributes()
                 break;
 
             default:
+                if (0 == typeName.Compare("Variant[]"))
```
Comment (eugeneko) [inline Source/Urho3D/AngelScript/ScriptInstance.cpp]: Why not `typeName == "Variant[]"`?

**Comment by eugeneko** [comment]
Thanks, I'll test your code later.

**Comment by orefkov** [comment]
About serialization of Variant[] and String[].
In documentation writed:
"After instantiation, the script object's public member variables that can be converted into Variant, and that don't begin with an underscore are automatically available as attributes of the ScriptInstance, and will be serialized."
In scripts we can convert to Variant  VariantMap, Variant[], String[].
But VariantMap is auto serialized in ScriptInstance, and Variant[] and String[] both not.
Imho it wrong.

**Comment by orefkov** [comment]
I was found and fix error when loaded scene in background.
In VectorToArray not possible get active script context, and it return null.

**Comment by eugeneko** [comment]
Merged. Thanks for patience.
BTW, I'm going to make some minor post-review changes.
The most important that Array::Swap IMO shall take inout& reference so you can't call it with const argument.

**Pull request merged**